pydrive2/auth.py

```diff
@@ -170,43 +81,119 @@ class GoogleAuth(ApiAttributeMixin):
-        else:
-            if self.settings is None:
-                self.settings = self.DEFAULT_SETTINGS
+
+        if settings is None and settings_file:
+            try:
+                settings = LoadSettingsFile(settings_file)
+            except SettingsError:
+                pass
+
+        self.settings = settings or self.DEFAULT_SETTINGS
+        ValidateSettings(self.settings)
+
+        self._service = None
+        self._client_config = None
+        self._oauth_type = None
+        self._flow = None
+        self._storage = None
+        self._credentials = None
+
+    # Lazy loading, read-only properties
+    @property
+    def service(self):
+        if not self._service:
+            self._service = build("drive", "v2", cache_discovery=False)
+        return self._service
+
+    @property
+    def client_config(self):
+        if not self._client_config:
+            self.LoadClientConfig()
+        return self._client_config
+
+    @property
+    def oauth_type(self):
+        if not self._oauth_type:
+            self.LoadClientConfig()
+        return self._oauth_type
+
+    @property
+    def flow(self):
+        if not self._flow:
+            self.GetFlow()
+        return self._flow
+
+    @property
+    def storage(self):
+        if not self.settings.get("save_credentials"):
+            return None
+
+        if not self._storage:
+            self._InitializeStoragesFromSettings()
+        return self._storage
+
+    @property
+    def credentials(self):
+        if not self._credentials:
+            if self.oauth_type in ("web", "installed"):
+                # try to load from backend if available
+                # credentials would auto-refresh if expired
+                if self.storage:
+                    try:
+                        self.LoadCredentials()
+                        return self._credentials
+                    except FileNotFoundError:
+                        pass
+
+                self.LocalWebserverAuth()
+
+            elif self.oauth_type == "service":
+                self.ServiceAuth()
             else:
-                ValidateSettings(self.settings)
-        self._storages = self._InitializeStoragesFromSettings()
-        # Only one (`file`) backend is supported now
-        self._default_storage = self._storages["file"]
+                raise InvalidConfigError(
+                    "Only web, installed, service oauth is supported"
+                )
+
+        return self._credentials
 
+    @property
+    def authorized_http(self):
+        # Ensure that a thread-safe, Authorized HTTP object is provided
+        # If HTTP object not specified, create or resuse an HTTP
+        # object from the thread local storage.
+        if not getattr(self.thread_local, "http", None):
+            self.thread_local.http = AuthorizedHttp(
+                self.credentials, http=self._build_http()
+            )
+
+        return self.thread_local.http
```

Comment:
@shcheklein Here's the lazy loading property that creates an `AuthorizedHttp` object & add to `self.thread_local` if it doesn't exist

Reply:
just to make sure - AuthorizedHttp handles refresh itself and updates the credentials object?

Reply:
Yup, `AuthorizedHttp` is part of the [google_auth_httplib2](https://github.com/googleapis/google-auth-library-python-httplib2). This library's goal is to facilitate the transition to the new google auth library